File: java/src/main/java/nobocoder/java/spelling/SpellChecker.java
package nobocoder.java.spelling;

import java.util.Collection;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

import nobocoder.java.util.Pair;

public interface SpellChecker {
	Optional<Collection<String>> suggestionsFor(String word);
	
	default Map<String, Collection<String>> suggestionsFor(Collection<String> words) {
		return words.stream().
			map(word -> new Pair<>(word, suggestionsFor(word))).
			filter(p -> p.getValue().isPresent()).
			collect(Collectors.toMap(Pair::getKey, p -> p.getValue().get()));
	}
}

File: scala/src/main/scala/nobocoder/scala/spelling/SpellChecker.scala
package nobocoder.scala.spelling

trait SpellChecker {
  def suggestionsFor(word: String): Option[Seq[String]]

  def suggestionsFor(words: Seq[String]): Map[String, Seq[String]] =
    words.
      map(word => (word, suggestionsFor(word))).
      filter(_._2.isDefined).
      toMap.
      mapValues(_.get)
}

trait NGramSpellChecker extends SpellChecker { this: DictionaryBuilder with NGram2WordBuilder with StringDistance  =>
  import NGram._

  def minSimilarity: Double

  lazy val ngram2Word = buildNGram2Word
  lazy val dictionary = buildDictionary

  def suggestionsFor(word: String): Option[Seq[String]] =
    if (dictionary contains word) None
    else Some {
      ngrams(word, 2).
        filter(ngram2Word.contains).
        flatMap(ngram2Word).
        distinct.
        map { candidate =>
          (candidate, distance(candidate, word))
        }.
        filter(_._2 >= minSimilarity).
        sortBy(-_._2).
        map(_._1)
    }
}

